### PR TITLE
re-add hydration support to React 18 errors

### PR DESCRIPTION
**Pull request by ztanner**
Re-lands hydration support for React 18 errors by reverting some things that were part of #65058. Annotated specific parts with comments.

### CODE REVIEW

**Comment by ztanner** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/vercel/next.js/69757?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#69757** <a href="https://app.graphite.dev/github/pr/vercel/next.js/69757?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#69484** <a href="https://app.graphite.dev/github/pr/vercel/next.js/69484?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#69733** <a href="https://app.graphite.dev/github/pr/vercel/next.js/69733?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#69619** <a href="https://app.graphite.dev/github/pr/vercel/next.js/69619?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#69610** <a href="https://app.graphite.dev/github/pr/vercel/next.js/69610?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#69564** <a href="https://app.graphite.dev/github/pr/vercel/next.js/69564?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#69494** <a href="https://app.graphite.dev/github/pr/vercel/next.js/69494?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#69492** <a href="https://app.graphite.dev/github/pr/vercel/next.js/69492?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#69515** <a href="https://app.graphite.dev/github/pr/vercel/next.js/69515?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#69512** <a href="https://app.graphite.dev/github/pr/vercel/next.js/69512?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ztanner and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/next/src/client/components/react-dev-overlay/internal/helpers/hydration-error-info.ts [hunk 1]
```diff
@@ -23,15 +23,38 @@ const htmlTagsWarnings = new Set([
...
+  'Warning: Expected server HTML to contain a matching text node for "%s" in <%s>.%s',
+  'Warning: Did not expect server HTML to contain the text node "%s" in <%s>.%s',
+])
+
+// Note: React 18 only
+const textMismatchWarning =
+  'Warning: Text content did not match. Server: "%s" Client: "%s"%s'
+
+const isTextMismatchWarning = (msg: NullableText) => textMismatchWarning === msg
+const isTextInTagsMismatchWarning = (msg: NullableText) =>
```
Comment (ztanner) [inline packages/next/src/client/components/react-dev-overlay/internal/helpers/hydration-error-info.ts:41]: Reverts: 
- https://github.com/vercel/next.js/pull/67903

File: packages/next/src/client/components/react-dev-overlay/internal/helpers/hydration-error-info.ts [hunk 2]
```diff
@@ -51,14 +74,18 @@ export const getReactHydrationDiffSegments = (msg: NullableText) => {
...
+    hydrationErrorState.warning = [msg, serverContent, clientContent]
     hydrationErrorState.componentStack = componentStack
     hydrationErrorState.serverContent = serverContent
     hydrationErrorState.clientContent = clientContent
+
+    return [
+      ...args,
+      // We tack on the hydration error message to the console.error message so that
+      // it matches the error we display in the redbox overlay
+      `\nSee more info here: https://nextjs.org/docs/messages/react-hydration-error`,
```
Comment (ztanner) [inline packages/next/src/client/components/react-dev-overlay/internal/helpers/hydration-error-info.ts:86]: Supports behavior that was lost from:
- https://github.com/vercel/next.js/pull/67011

File: test/development/acceptance/hydration-error.test.ts [hunk 3]
```diff
@@ -38,7 +37,7 @@ describe.skip('Error overlay for hydration errors (React 18)', () => {
     await session.assertHasRedbox()
 
     expect(await session.getRedboxDescription()).toMatchInlineSnapshot(`
-      "Error: Text content does not match server-rendered HTML.
+      "Text content does not match server-rendered HTML.
```
Comment (ztanner) [inline test/development/acceptance/hydration-error.test.ts:40]: Slight change in behavior but showing `Error:` in an error overlay feels a bit redundant

File: packages/next/src/client/components/react-dev-overlay/internal/container/Errors.tsx [hunk 4]
```diff
@@ -307,7 +307,9 @@ export function Errors({
               {/* If there's hydration warning, skip displaying the error name */}
               {hydrationWarning ? '' : error.name + ': '}
               <HotlinkedText
-                text={hydrationWarning || error.message}
+                text={
+                  isAppDir ? hydrationWarning || error.message : error.message
```
Comment (ztanner) [inline packages/next/src/client/components/react-dev-overlay/internal/container/Errors.tsx:311]: Reverts the `Errors.tsx` changes from:
- https://github.com/vercel/next.js/pull/65276

to match the React 18 behavior when not in app dir

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 81ceea6c50802ba81c1b5785601dae2ba20702c4

`TURBOPACK=1 pnpm test test/integration/auto-export/test/index.test.js` (turbopack)
- Auto Export > dev > should include error link when hydration error does occur

<details>
<summary>Expand output</summary>

  ● Auto Export › dev › should include error link when hydration error does occur

    expect(received).toEqual(expected) // deep equality

    Expected: ArrayContaining [{"message": StringContaining "https://react.dev/link/hydration-mismatch", "source": "error"}]
    Received: [{"message": "%cDownload the React DevTools for a better development experience: https://reactjs.org/link/react-devtools font-weight:bold", "source": "info"}, {"message": "[HMR] connected", "source": "log"}, {"message": "[Fast Refresh] rebuilding", "source": "log"}, {"message": "Warning: Text content did not match. Server: \"%s\" Client: \"%s\"%s /[post]/[cmnt] hydration error·
        at p
        at Page

      at App (../http:/localhost:37095/_next/static/chunks/%5Broot%20of%20the%20server%5D__5a9fa8._.js:20970:1)
      at PathnameContextProviderAdapter (../http:/localhost:37095/_next/static/chunks/%5Broot%20of%20the%20server%5D__5a9fa8._.js:8831:11)
      at ErrorBoundary (../http:/localhost:37095/_next/static/chunks/%5Broot%20of%20the%20server%5D__5a9fa8._.js:18047:9)
      at ReactDevOverlay (../http:/localhost:37095/_next/static/chunks/%5Broot%20of%20the%20server%5D__5a9fa8._.js:18351:11)
      at Container (../http:/localhost:37095/_next/static/chunks/%5Broot%20of%20the%20server%5D__5a9fa8._.js:18676:1)
      at AppContainer (../http:/localhost:37095/_next/static/chunks/%5Broot%20of%20the%20server%5D__5a9fa8._.js:18787:11)
      at Root (../http:/localhost:37095/_next/static/chunks/%5Broot%20of%20the%20server%5D__5a9fa8._.js:18960:11)·
      See more info here: https://nextjs.org/docs/messages/react-hydration-error", "source": "error"}, {"message": "Warning: An error occurred during hydration. The server HTML was replaced with client content in <%s>. div", "source": "error"}, {"message": "Text content does not match server-rendered HTML.", "source": "error"}, {"message": "There was an error while hydrating. Because the error happened outside of a Suspense boundary, the entire root will switch to client rendering.", "source": "error"}, {"message": "connected to ws at ws://localhost:37095/_next/webpack-hmr", "source": "log"}, {"message": "Next.js page already hydrated", "source": "log"}, {"message": "received ws message {\"action\":\"appIsrManifest\",\"data\":{}}", "source": "log"}, …]
      at Object.toEqual (integration/auto-export/test/index.test.js:98:20)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: test/development/acceptance/hydration-error-react-19.test.ts [hunk 5]
```diff
@@ -4,7 +4,8 @@ import { FileRef, nextTestSetup } from 'e2e-utils'
 import { outdent } from 'outdent'
 import path from 'path'
 
-describe('Error overlay for hydration errors (React 19)', () => {
+// TODO: Enable once React 19 support is added to pages.
+describe.skip('Error overlay for hydration errors (React 19)', () => {
```
Comment (ztanner) [inline test/development/acceptance/hydration-error-react-19.test.ts:8]: This can't run because we throw in pages if 18 is used

**Review by huozhi — APPROVED** [review]

**Pull request merged**